sw/device/lib/testing/dif/mock_dif_uart_test.cc

```diff
@@ -0,0 +1,228 @@
+
+  EXPECT_TRUE(dif_uart_configure(&dif_uart(), &dif_uart_config()));
+}
+
+/**
+ * Difference from ConfigureSuccessParityEven, parity_type=odd.
+ */
+TEST_F(DifUartTestUartInitialised, ConfigureSuccessParityOdd) {
+  dif_uart_config_t *config = &dif_uart_config();
+  config->parity_enable = kDifUartEnable;
+  config->parity = kDifUartParityOdd;
+
+  EXPECT_WRITE32(UART_CTRL_REG_OFFSET, 0x000100C3);
+  EXPECT_WRITE32(UART_FIFO_CTRL_REG_OFFSET, 0x00000003);
+  EXPECT_WRITE32(UART_INTR_ENABLE_REG_OFFSET, 0x00000000);
+
+  EXPECT_TRUE(dif_uart_configure(&dif_uart(), &dif_uart_config()));
+}
+
+// dif_uart_watermark_rx_set() tests
+
+/**
+ * TODO
+ */
+TEST_F(DifUartTestUartInitialised, WatermarkRxSetSuccess) {
+  EXPECT_READ32(UART_FIFO_CTRL_REG_OFFSET, 0x00000000);
+  EXPECT_WRITE32(UART_FIFO_CTRL_REG_OFFSET, 0x00000000);
+
+  EXPECT_TRUE(dif_uart_watermark_rx_set(&dif_uart(), kDifUartWatermarkByte1));
+
+  EXPECT_READ32(UART_FIFO_CTRL_REG_OFFSET, 0x00000000);
+  EXPECT_WRITE32(UART_FIFO_CTRL_REG_OFFSET, 0x00000004);
+
+  EXPECT_TRUE(dif_uart_watermark_rx_set(&dif_uart(), kDifUartWatermarkByte4));
+
+  EXPECT_READ32(UART_FIFO_CTRL_REG_OFFSET, 0x00000000);
+  EXPECT_WRITE32(UART_FIFO_CTRL_REG_OFFSET, 0x00000008);
+
+  EXPECT_TRUE(dif_uart_watermark_rx_set(&dif_uart(), kDifUartWatermarkByte8));
+
+  EXPECT_READ32(UART_FIFO_CTRL_REG_OFFSET, 0x00000000);
+  EXPECT_WRITE32(UART_FIFO_CTRL_REG_OFFSET, 0x0000000C);
+
+  EXPECT_TRUE(dif_uart_watermark_rx_set(&dif_uart(), kDifUartWatermarkByte16));
+
+  EXPECT_READ32(UART_FIFO_CTRL_REG_OFFSET, 0x00000000);
+  EXPECT_WRITE32(UART_FIFO_CTRL_REG_OFFSET, 0x00000010);
+
+  EXPECT_TRUE(dif_uart_watermark_rx_set(&dif_uart(), kDifUartWatermarkByte30));
+}
+
+// dif_uart_watermark_tx_set() tests
+
+/**
+ * TODO
+ */
+TEST_F(DifUartTestUartInitialised, WatermarkTxSetSuccess) {
+  EXPECT_READ32(UART_FIFO_CTRL_REG_OFFSET, 0x00000000);
+  EXPECT_WRITE32(UART_FIFO_CTRL_REG_OFFSET, 0x00000000);
+
+  EXPECT_TRUE(dif_uart_watermark_tx_set(&dif_uart(), kDifUartWatermarkByte1));
+
+  EXPECT_READ32(UART_FIFO_CTRL_REG_OFFSET, 0x00000000);
+  EXPECT_WRITE32(UART_FIFO_CTRL_REG_OFFSET, 0x00000020);
+
+  EXPECT_TRUE(dif_uart_watermark_tx_set(&dif_uart(), kDifUartWatermarkByte4));
+
+  EXPECT_READ32(UART_FIFO_CTRL_REG_OFFSET, 0x00000000);
+  EXPECT_WRITE32(UART_FIFO_CTRL_REG_OFFSET, 0x00000040);
+
+  EXPECT_TRUE(dif_uart_watermark_tx_set(&dif_uart(), kDifUartWatermarkByte8));
+
+  EXPECT_READ32(UART_FIFO_CTRL_REG_OFFSET, 0x00000000);
+  EXPECT_WRITE32(UART_FIFO_CTRL_REG_OFFSET, 0x00000060);
+
+  EXPECT_TRUE(dif_uart_watermark_tx_set(&dif_uart(), kDifUartWatermarkByte16));
+}
+
+// dif_uart_bytes_send() tests
+
+/**
+ * TODO
+ */
+TEST_F(DifUartTestUartInitialised, BytesSendTxFifoEmptySuccess) {
+  uint8_t data[32] = {
+      'a', 'a', 'a', 'a', 'a', 'a', 'a', 'a', 'b', 'b', 'b',
+      'b', 'b', 'b', 'b', 'b', 'c', 'c', 'c', 'c', 'c', 'c',
+      'c', 'c', 'd', 'd', 'd', 'd', 'd', 'd', 'd', 'd',
+  };
```

Comment:
Consider making this a `constexpr` global of type `std::array` so that you can re-use it in other tests, and so that you can use range-for and size() on it.